### PR TITLE
Issue/4341 payment tutorial

### LINKED ISSUE #4341: [Card Present Payments] Implement tutorial
Design - vaw3DvewbUwuQnPXpdNGGH-fi-121%3A0

### PR DESCRIPTION
**Pull request by nbradbury**
Closes #4341 by implementing the card reader tutorial following the first successful connection.

tut

**Individual screenshots:**

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Review by Garance91540 — APPROVED** [review]
Looking good!

**Comment by malinajirka** [comment]
Thanks for working on this @nbradbury: Just a note about the "automatic reconnection" step - it's [not implemented yet](https://github.com/woocommerce/woocommerce-android/projects/28), so I'd consider removing this step for now. Wdyt?

**Comment by nbradbury** [comment]
> Just a note about the "automatic reconnection" step - it's [not implemented yet](https://github.com/woocommerce/woocommerce-android/projects/28), so I'd consider removing this step for now. Wdyt?

@malinajirka Ah, thanks for pointing that out. I'll simply comment out that step and we can un-comment it when reconnection is supported.

**Comment by malinajirka** [comment]
@Garance91540 The "Your reader can stay on for approximately 3 hours .." message based on facts or just an example for the designs? I've heard the reader can process +- 600 payments, but I haven't heard anything about 3 hours - I'd expect it to last for days when it's not accepting any payments.

**Comment by nbradbury** [comment]
> I've heard the reader can process +- 600 payments, but I haven't heard anything about 3 hours - I'd expect it to last for days when it's not accepting any payments.

@Garance91540 @malinajirka Good catch, all I see in [the product sheet](https://www.bbpos.com/chipper-2x-bt/) is "Above 800 EMV transactions, or above 5000 magnetic card swipe transactions, or above 1000 EMV contactless transactions for one full charge."

**Comment by kidinov** [comment]
@nbradbury @malinajirka Should we consider not merging that before the cut-off? The change is pretty big so it's might a bit risky to make it at the last moment?

**Comment by malinajirka** [comment]
> Should we consider not merging that before the cut-off? The change is pretty big so it's might a bit risky to make it at the last moment?

I believe we are still in the development cycle. We have two more weeks of testing before this goes to production. So I personally wouldn't worry about it. Having said that there are still some open questions and the review will be time consuming, so we might not merge it in time anyway, we'll see.

**Comment by nbradbury** [comment]
> Should we consider not merging that before the cut-off? The change is pretty big so it's might a bit risky to make it at the last moment?

As @malinajirka mentioned, I think we should hold off merging until the text for the third page is clarified. But starting a code review now is fine.

**Comment by nbradbury** [comment]
**Update:** As discussed on Slack, we noticed that [the Chipper web site](https://www.bbpos.com/chipper-2x-bt/) says "Charging: Via Micro USB, 3 hours for fully charged,” which must be where the 3 hours came from. We decided to [change](https://github.com/woocommerce/woocommerce-android/pull/4372/commits/0294203e660f91066812bfe7729c67beb269fd77) our wording to "Your reader takes about three hours to fully charge."

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/tutorial/CardReaderTutorialDialogFragment.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        binding.viewPagerIndicator.setupFromViewPager(binding.viewPager)
+        binding.viewPager.addOnPageChangeListener(object : ViewPager.SimpleOnPageChangeListener() {
+            override fun onPageSelected(position: Int) {
+                val lastPosition = CardReaderTutorialViewPager.NUM_PAGES - 1
+                binding.buttonSkip.setText(if (position == lastPosition) R.string.close else R.string.skip)
+            }
+        })
+
+        binding.buttonSkip.setOnClickListener {
+            navigateBackWithResult(KEY_READER_TUTORIAL_RESULT, true)
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/tutorial/CardReaderTutorialDialogFragment.kt]: It looks liek that we always send "true" back and this boolean not used. Should we use `navigateBackWithNotice` with notice then?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/tutorial/CardReaderTutorialViewPagerItemFragment.kt [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import com.woocommerce.android.R
+import com.woocommerce.android.databinding.FragmentCardReaderTutorialViewpagerItemBinding
+import com.woocommerce.android.extensions.hide
+import org.wordpress.android.util.DisplayUtils
+
+/**
+ * Displays a single image and text label in the card reader tutorial view pager
+ */
+class CardReaderTutorialViewPagerItemFragment : Fragment(R.layout.fragment_card_reader_tutorial_viewpager_item) {
+    companion object {
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/tutorial/CardReaderTutorialViewPagerItemFragment.kt]: Bad that detekt doesn't handle this (maybe in the future versions), but accroding to the convention companion objects should go to the end of the class

https://kotlinlang.org/docs/coding-conventions.html#class-layout

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/tutorial/CardReaderTutorialViewPager.kt [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import android.content.Context
+import android.util.AttributeSet
+import androidx.fragment.app.Fragment
+import androidx.fragment.app.FragmentManager
+import androidx.fragment.app.FragmentPagerAdapter
+import com.woocommerce.android.R
+import com.woocommerce.android.widgets.WCViewPager
+
+class CardReaderTutorialViewPager : WCViewPager {
+    companion object {
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/tutorial/CardReaderTutorialViewPager.kt]: Same comment here aboyt companion objects

**Review by kidinov — COMMENTED** [review]
Looks good! Please check my comments

**Comment by kidinov** [comment]
Some visual concerns:
* Dots look smaller than on the design
* Something wrong with the images. Check the lines inside of the reader for instance
* Selected dot is white, but on the design it's #B17FD4

screen
screen1

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/connect/CardReaderConnectViewModel.kt [hunk 4]
```diff
@@ -268,6 +271,16 @@ class CardReaderConnectViewModel @Inject constructor(
...
+        // show the tutorial if this is the first time the user has connected a reader, otherwise we're done
+        if (appPrefs.getShowCardReaderConnectedTutorial()) {
+            triggerEvent(ShowCardReaderTutorial)
+        } else {
+            exitFlow(connected = true)
+        }
+    }
+
+    fun onTutorialClosed() {
+        appPrefs.setShowCardReaderConnectedTutorial(false)
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/connect/CardReaderConnectViewModel.kt]: Probably not really important, but would it make sense to make it right after we show the tutorial? (line 276) Currently, if I close the app during the tutorial is shown, then the next time I connect to a reader I'll see the tutorial again

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/tutorial/CardReaderTutorialDialogFragment.kt (hunk 1 above)
Comment (nbradbury) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/tutorial/CardReaderTutorialDialogFragment.kt]: Ah, thanks - I wasn't aware of `navigateBackWithNotice`. Change made in 20776f4.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/tutorial/CardReaderTutorialViewPagerItemFragment.kt (hunk 2 above)
Comment (nbradbury) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/tutorial/CardReaderTutorialViewPagerItemFragment.kt]: I wasn't aware of this, either. Change made in 14f3732.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/connect/CardReaderConnectViewModel.kt (hunk 4 above)
Comment (nbradbury) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/connect/CardReaderConnectViewModel.kt]: Changed in 371b567.

**Comment by nbradbury** [comment]
> * Selected dot is white, but on the design it's #B17FD4

@kidinov I made that change, but the difference in selected/unselected colors in very subtle.

indicator

We're using the same indicator as the login prologue, and there it's a much more noticeable light gray. My vote is to keep the light gray for both screens. @woocommerce/mobile-designers, WDYT?

Screen Shot 2021-07-12 at 7 52 19 AM

**Comment by nbradbury** [comment]
> * Dots look smaller than on the design

@kidinov Fixed in 5e7a4a6.

**Comment by nbradbury** [comment]
> * Something wrong with the images. Check the lines inside of the reader for instance

@kidinov @Garance91540 Hmm, looks like something may be off with the original images. They look fine in Figma, but when I copy the SVG from Figma and view it in Chrome, it looks the same as in the app.

Screen Shot 2021-07-12 at 8 49 32 AM

I see similar issues with the "card reader available" image, which was added prior to the tutorial.

Screen Shot 2021-07-12 at 10 01 51 AM

**Comment by ghost** [comment]
@nbradbury My vote is to keep the light gray indicator for both screens too.

Thanks for raising the issue with the SVG. I'll take a look and send them through to you

**Comment by nbradbury** [comment]
@kidinov WDYT about merging this now, despite a few visual glitches? The same glitches happen outside of this PR as well, so once we have the updated images I could submit a PR with all the updates. If the problems were glaring I wouldn't suggest this, but since they went unnoticed for a while I think we're safe, and it would be nice not to have to keep this synched with `develop`!

**Comment by Garance91540** [comment]
Hi @nbradbury , 

> The same glitches happen outside of this PR as well

Could you point me to those places? we definitely should fix that asap!

> we have the updated images I could submit a PR with all the updates

I don't think we have new images to submit unfortunately! they are displayed correctly on Figma, so I think the problem is in the export feature... @anitaa1990 I know you did some magic in the past on that matter, do you have any insights on how to fix this export asset issue?

> WDYT about merging this now, despite a few visual glitches

I'd prefer if we could prioritize fixing this bug before releasing it if possible. (Letting those kind of visual glitches creep up incrementally can really hurt the overall perception of our product). Let me know if there's something else I could try that would help export those assets correctly.

**Comment by nbradbury** [comment]
@Garance91540 In a Slack DM @adamzelinski sent me two SVGs of just the reader and the card, and the glitch in them was solved. We just need updated versions of these images:

* Card reader available
* Card insert
* Card sleep
* Card charge

I tried manually correcting the images, but hand-editing SVGs is apparently out of range for me :)

**Comment by nbradbury** [comment]
> Could you point me to those places?

@Garance91540 As [mentioned above](https://github.com/woocommerce/woocommerce-android/pull/4372#issuecomment-878252648), the "card reader available" image shows the same problem.

**Comment by anitaa1990** [comment]
@Garance91540 @nbradbury I'm not sure if this would solve your problem but I had previously exported the PDF version of the image from Figma, then used [this online tool](https://www.zamzar.com/convert/pdf-to-svg/) to convert the PDF to svg. Then used [another online tool](https://svg2vector.com/) to convert the svg to vector drawable. 

Someone had mentioned iOS platform exports the images to pdf from Figma and it seemed to work so I had followed that route. Not sure if it would be applicable here but sharing in case it' useful :) 

More details in this P2 post: `wp.me/p91TBi-4Gb#comment-3821`

**Comment by nbradbury** [comment]
> I'm not sure if this would solve your problem but I had previously exported the PDF version of the image from Figma...

Thanks @anitaa1990! That helped in a some cases, but not all of them. In the end I did a whole lot of manual tweaking of the vector itself, which was time consuming but at least it worked :)

**Comment by nbradbury** [comment]
@kidinov @Garance91540 I spent the better part of my day manually editing the SVGs/vectors, and I think it's all looking good now.

light

dark

**Comment by Garance91540** [comment]
Hi @nbradbury , looking good! Thanks for fixing those. 

I took a closer look at the assets that were causing issues, and it seems like they all were related with strokes. So in the future when sharing svg assets, I'll try to only share filled-in components vs stroke-heavy ones and hopefully that helps!

I just sent you the "card reader available" svg, let me know if you need anything else!

**Comment by kidinov** [comment]
@nbradbury Tremendous job with the pictures!

Have you considered to change the layout `fragment_card_reader_tutorial_viewpager_item` to

```
<?xml version="1.0" encoding="utf-8"?>
<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:app="http://schemas.android.com/apk/res-auto"
    xmlns:tools="http://schemas.android.com/tools"
    android:layout_width="match_parent"
    android:layout_height="match_parent"
    android:gravity="center"
    android:orientation="vertical">

    <com.google.android.material.textview.MaterialTextView
        android:id="@+id/labelTextView"
        style="@style/Woo.TextView.Subtitle2"
        android:layout_width="wrap_content"
        android:layout_height="wrap_content"
        android:gravity="center"
        app:layout_constraintEnd_toEndOf="parent"
        app:layout_constraintStart_toStartOf="parent"
        app:layout_constraintTop_toTopOf="parent"
        tools:layout_width="200dp"
        tools:text="@string/card_reader_tutorial_connected_label" />

    <ImageView
        android:id="@+id/imageView"
        android:layout_width="wrap_content"
        android:layout_height="wrap_content"
        android:importantForAccessibility="no"
        app:layout_constraintBottom_toTopOf="@+id/detailTextView"
        app:layout_constraintEnd_toEndOf="parent"
        app:layout_constraintStart_toStartOf="parent"
        app:layout_constraintTop_toBottomOf="@+id/labelTextView"
        tools:src="@drawable/img_wc_ship_payment" />

    <com.google.android.material.textview.MaterialTextView
        android:id="@+id/detailTextView"
        style="@style/Woo.TextView.Body2"
        android:layout_width="wrap_content"
        android:layout_height="wrap_content"
        android:gravity="center"
        app:layout_constraintBottom_toBottomOf="parent"
        app:layout_constraintEnd_toEndOf="parent"
        app:layout_constraintStart_toStartOf="parent"
        tools:layout_width="200dp"
        tools:text="@string/card_reader_tutorial_connected_detail" />
</androidx.constraintlayout.widget.ConstraintLayout>

```

Also, in `dialog_card_reader_tutorial`

```
        <com.woocommerce.android.ui.prefs.cardreader.tutorial.CardReaderTutorialViewPager
            android:id="@+id/viewPager"
            android:layout_width="wrap_content"
            android:layout_height="0dp"
```

So the text is aligned on different slides?

https://user-images.githubusercontent.com/4923871/125759510-58305470-fb67-4460-b8c6-3c43925e73a6.mp4

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/connect/CardReaderConnectViewModel.kt [hunk 5]
```diff
@@ -283,10 +284,22 @@ class CardReaderConnectViewModel @Inject constructor(
...
+            exitFlow(connected = true)
+        }
+    }
+
+    fun onTutorialClosed() {
         exitFlow(connected = true)
     }
 
     private fun exitFlow(connected: Boolean) {
+        appPrefs.setShowCardReaderConnectedTutorial(true) // TODO remove
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/connect/CardReaderConnectViewModel.kt]: Don't forget to remove it before merging! =)

**Comment by nbradbury** [comment]
@kidinov I made the recommended layout changes in aa30d87 and removed the TODO item in 9601df7.

**Review by kidinov — APPROVED** [review]
Great job 👍

**Pull request merged**